Additional Tests and Error Handling for XML

s9latimm opened the pull request:
resolves se-sic/VaRA#770

s9latimm committed:
Add errors.

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/Error.h
M	include/vara/Feature/FeatureModelTransaction.h
M	lib/Feature/Constraint.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/resources/CMakeLists.txt
A	unittests/resources/xml/error_mismatch_parent_child.xml
A	unittests/resources/xml/error_missing_child.xml
A	unittests/resources/xml/error_missing_exclude.xml
A	unittests/resources/xml/error_missing_implication.xml
A	unittests/resources/xml/error_missing_parent.xml
A	unittests/resources/xml/error_root_root.xml
A	unittests/resources/xml/test_locations.xml

vulder commented on unittests/Feature/FeatureModelParser.cpp:
```diff
@@ -53,6 +53,75 @@ TEST(FeatureModelParser, outOfOrder) {
   EXPECT_TRUE(P.buildFeatureModel());
 }
 
+// TODO: Detect error while parsing
```
What's the idea behind this?

vulder commented on unittests/Feature/FeatureModelParser.cpp:
```diff
@@ -53,6 +53,75 @@ TEST(FeatureModelParser, outOfOrder) {
+// }
+
+TEST(FeatureModelParser, errorMissingChild) {
+  auto FS = llvm::MemoryBuffer::getFileAsStream(
+      getTestResource("error_missing_child.xml"));
+  ASSERT_TRUE(FS);
+
+  auto P = FeatureModelXmlParser(FS.get()->getBuffer().str());
+
+  EXPECT_TRUE(P.verifyFeatureModel());
+  EXPECT_FALSE(P.buildFeatureModel());
```
To simplify things, we should move this into a method that does the initial checks and returns P, should some test want to test extra stuff. With this, a test is only putting a string into the new method.

vulder commented on unittests/Feature/FeatureModelParser.cpp:
```diff
@@ -53,6 +53,75 @@ TEST(FeatureModelParser, outOfOrder) {
+//   EXPECT_FALSE(P.buildFeatureModel());
+// }
+
+TEST(FeatureModelParser, errorMissingChild) {
```
For the tests in general, are they ensuring enough? It looks a bit strange that we only verify/build an then accept. It could be that the tests check to general maybe, we should talk about this.

vulder committed:
Merge branch 'vara-dev' into f-XMLTests

s9latimm committed:
Simplify tests

M	unittests/Feature/FeatureModelParser.cpp

s9latimm committed:
Detect ambiguous edges

M	include/vara/Feature/FeatureModelBuilder.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/resources/xml/error_mismatch_parent_child.xml

s9latimm committed:
Revert tested model

M	unittests/resources/xml/error_mismatch_parent_child.xml

s9latimm committed:
Use iterator

M	include/vara/Feature/FeatureModelBuilder.h

vulder commented on include/vara/Feature/FeatureModelBuilder.h:
```diff
@@ -57,8 +57,8 @@ class FeatureModelBuilder {
   std::optional<std::string> getParentName(const std::string &FeatureName) {
-    if (Parents.find(FeatureName) != Parents.end()) {
-      return Parents[FeatureName];
+    if (auto P = Parents.find(FeatureName); P != Parents.end()) {
```
```suggestion
    if (const auto *P = Parents.find(FeatureName); P != Parents.end()) {
```

vulder committed:
Apply suggestions from code review

vulder commented on include/vara/Feature/FeatureModelBuilder.h:
```diff
@@ -51,10 +51,18 @@ class FeatureModelBuilder {
   }
 
+  std::optional<std::string> getParentName(const std::string &FeatureName) {
+    if (const auto *P = Parents.find(FeatureName); P != Parents.end()) {
```
```suggestion
    if (const auto P = Parents.find(FeatureName); P != Parents.end()) {
```

vulder committed:
Apply suggestions from code review

vulder commented on include/vara/Feature/FeatureModelBuilder.h:
```diff
@@ -51,10 +51,18 @@ class FeatureModelBuilder {
     return this;
   }
 
+  std::optional<std::string> getParentName(const std::string &FeatureName) {
```
```suggestion
  std::optional<std::string> getParentName(const std::string &FeatureName) const {
```

vulder committed:
Update include/vara/Feature/FeatureModelBuilder.h

vulder committed:
Update include/vara/Feature/FeatureModelBuilder.h

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
